File: packages/pangle/ios/HybridPangleInterstitialAd.swift
import Foundation
import NitroModules
import PAGAdSDK

final class HybridPangleInterstitialAd:
  HybridPangleInterstitialAdSpec, @unchecked Sendable {
  fileprivate let events = PangleAdEventEmitter()
  private let placementId: String
  fileprivate var ad: PAGLInterstitialAd?
  fileprivate let showTimeout = PangleShowTimeoutController()
  private lazy var delegate = PangleInterstitialDelegate(owner: self)

  init(placementId: String) {
    self.placementId = placementId
  }

  func load() throws {
    DispatchQueue.main.async {
      let request = PAGInterstitialRequest()
      PAGLInterstitialAd.load(withSlotID: self.placementId, request: request) {
        interstitialAd, error in
        DispatchQueue.main.async {
          guard let interstitialAd else {
            self.events.emit(.error, error: error ?? pangleError("Pangle interstitial ad failed to load"))
            return
          }
          interstitialAd.delegate = self.delegate
          self.ad = interstitialAd
          self.events.emit(.loaded)
        }
      }
    }
  }

  func show() throws -> Promise<Void> {
    Promise.async {
      do {
        let viewController = try await currentViewController()
        try await withCheckedThrowingContinuation {
          (continuation: CheckedContinuation<Void, Error>) in
          DispatchQueue.main.async {
            guard let ad = self.ad else {
              continuation.resume(
                throwing: pangleError("Pangle interstitial ad is not ready")
              )
              return
            }
            self.showTimeout.arm { [weak self] in
              guard let self else { return }
              self.ad = nil
              self.events.emit(.error, payload: PangleAdEventPayload(
                errorCode: nil,
                errorMessage: "Pangle ad show() timed out without a response from the SDK"
              ))
            }
            ad.present(fromRootViewController: viewController)
            continuation.resume()
          }
        }
      } catch {
        self.events.emit(.error, error: error)
        throw error
      }
    }
  }

  func addAdEventListener(
    eventType: PangleAdEventType,
    listener: @escaping (PangleAdEventPayload?) -> Void
  ) throws -> Double {
    events.add(eventType: eventType, listener: listener)
  }

  func removeAdEventListener(subscriptionId: Double) throws {
    events.remove(subscriptionId: subscriptionId)
  }

  func removeAllListeners() throws {
    events.removeAll()
  }
}

private final class PangleInterstitialDelegate: NSObject, PAGLInterstitialAdDelegate {
  private weak var owner: HybridPangleInterstitialAd?

  init(owner: HybridPangleInterstitialAd) {
    self.owner = owner
  }

  func adDidShow(_ ad: any PAGAdProtocol) {
    owner?.showTimeout.disarm()
    owner?.events.emit(.opened)
  }

  func adDidClick(_ ad: any PAGAdProtocol) {
    owner?.events.emit(.clicked)
  }

  func adDidDismiss(_ ad: any PAGAdProtocol) {
    owner?.showTimeout.disarm()
    owner?.ad = nil
    owner?.events.emit(.closed)
  }

  func adDidShowFail(_ ad: any PAGAdProtocol, error: Error) {
    owner?.showTimeout.disarm()
    owner?.ad = nil
    owner?.events.emit(.error, error: error)
  }
}

File: packages/pangle/android/src/main/java/com/margelo/nitro/pangleads/HybridPangleInterstitialAd.kt
package com.margelo.nitro.pangleads

import android.os.Handler
import android.os.Looper
import androidx.annotation.Keep
import com.bytedance.sdk.openadsdk.api.interstitial.PAGInterstitialAd
import com.bytedance.sdk.openadsdk.api.interstitial.PAGInterstitialAdInteractionListener
import com.bytedance.sdk.openadsdk.api.interstitial.PAGInterstitialAdLoadListener
import com.bytedance.sdk.openadsdk.api.interstitial.PAGInterstitialRequest
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.bridge.ReactApplicationContext
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

@DoNotStrip
@Keep
class HybridPangleInterstitialAd(
  private val context: ReactApplicationContext,
  private val placementId: String,
) : HybridPangleInterstitialAdSpec(), PAGInterstitialAdInteractionListener {
  private val events = PangleAdEventEmitter()
  private val mainHandler = Handler(Looper.getMainLooper())
  private var ad: PAGInterstitialAd? = null
  private var showTimeout: Runnable? = null

  override fun load() {
    context.runOnUiQueueThread {
      PAGInterstitialAd.loadAd(
        placementId,
        PAGInterstitialRequest(),
        object : PAGInterstitialAdLoadListener {
          override fun onError(code: Int, message: String) {
            mainHandler.post {
              events.emit(
                PangleAdEventType.ERROR,
                pangleException("load", code, message).toPangleEventPayload(),
              )
            }
          }

          override fun onAdLoaded(interstitialAd: PAGInterstitialAd) {
            mainHandler.post {
              interstitialAd.setAdInteractionListener(this@HybridPangleInterstitialAd)
              ad = interstitialAd
              events.emit(PangleAdEventType.LOADED)
            }
          }
        },
      )
    }
  }

  override fun show(): Promise<Unit> = Promise.async {
    withContext(Dispatchers.Main) {
      val activity = context.requireCurrentActivity()
      val current = ad ?: error("Pangle interstitial ad is not ready")
      armShowTimeout()
      current.show(activity)
    }
  }

  override fun addAdEventListener(
    eventType: PangleAdEventType,
    listener: (payload: PangleAdEventPayload?) -> Unit,
  ) = events.add(eventType, listener)

  override fun removeAdEventListener(subscriptionId: Double) {
    events.remove(subscriptionId)
  }

  override fun removeAllListeners() {
    events.removeAll()
  }

  override fun onAdShowed() {
    disarmShowTimeout()
    events.emit(PangleAdEventType.OPENED)
  }

  override fun onAdClicked() {
    events.emit(PangleAdEventType.CLICKED)
  }

  override fun onAdDismissed() {
    disarmShowTimeout()
    ad = null
    events.emit(PangleAdEventType.CLOSED)
  }

  private fun armShowTimeout() {
    disarmShowTimeout()
    val runnable = Runnable {
      showTimeout = null
      ad = null
      events.emit(PangleAdEventType.ERROR, pangleShowTimeoutPayload())
    }
    showTimeout = runnable
    mainHandler.postDelayed(runnable, PANGLE_SHOW_TIMEOUT_MS)
  }

  private fun disarmShowTimeout() {
    showTimeout?.let(mainHandler::removeCallbacks)
    showTimeout = null
  }
}
